File: android/app/src/main/java/com/lightningkite/kwifttemplate/views/MapDemoVG.shared.kt
package com.lightningkite.kwifttemplate.views

import android.view.View
import com.lightningkite.kwift.delay
import com.lightningkite.kwift.location.GeoAddress
import com.lightningkite.kwift.location.coordinate
import com.lightningkite.kwift.location.geocode
import com.lightningkite.kwift.location.oneLine
import com.lightningkite.kwift.location.GeoCoordinate
import com.lightningkite.kwift.maps.PlacesAutocomplete
import com.lightningkite.kwift.maps.bindSelect
import com.lightningkite.kwift.observables.binding.bind
import com.lightningkite.kwift.observables.binding.bindString
import com.lightningkite.kwift.observables.StandardObservableProperty
import com.lightningkite.kwift.observables.addAndRunWeak
import com.lightningkite.kwift.observables.transformed
import com.lightningkite.kwift.views.ViewDependency
import com.lightningkite.kwift.views.onClick
import com.lightningkite.kwift.views.ViewGenerator
import com.lightningkite.kwifttemplate.layouts.ControlsDemoXml
import com.lightningkite.kwifttemplate.layouts.MapDemoXml
import com.lightningkite.kwifttemplate.layouts.RowTextXml

class MapDemoVG() : ViewGenerator() {
    override val title: String get() = "Map Demo"

    val text: StandardObservableProperty<String> = StandardObservableProperty("")
    val position: StandardObservableProperty<GeoCoordinate?> = StandardObservableProperty(null)
    val options: StandardObservableProperty<List<GeoAddress>> = StandardObservableProperty(listOf())

    override fun generate(dependency: ViewDependency): View {
        val xml = MapDemoXml()
        val view = xml.setup(dependency)

        xml.map.bindSelect(dependency, position)
        xml.select.bindString(text)
        xml.options.bind(options, GeoAddress()) { obs ->
            val xml = RowTextXml()
            val view = xml.setup(dependency)
            xml.label.bindString(obs.transformed { it -> it.oneLine() })
            xml.xmlRoot.onClick {
                this.position.value = obs.value.coordinate
            }
            return@bind view
        }

        text.addAndRunWeak(this){ self, value ->
            delay(1000){
                if(self.text.value == value && value.isNotBlank()){
                    dependency.geocode(value) { addresses ->
                        self.options.value = addresses
                    }
                }
            }
        }

        return view
    }
}

File: ios/KhrysalisTemplate/com/lightningkite/khrysalistemplate/vg/MapDemoVG.shared.swift
//Package: com.lightningkite.khrysalistemplate.vg
//Converted using Khrysalis2

import Foundation
import Khrysalis
import RxSwift
import RxRelay
import KhrysalisMaps
import KhrysalisBluetooth



public class MapDemoVG: ViewGenerator {
    
    
    override public var title: String {
        get {
            return "Map Demo"
        }
    }
    public var text: StandardObservableProperty<String>
    public var position: StandardObservableProperty<GeoCoordinate?>
    public var options: StandardObservableProperty<Array<GeoAddress>>
    
    override public func generate(dependency: ViewDependency) -> View {
        var xml = MapDemoXml()
        var view = xml.setup(dependency)
        xml.map.bindSelect(dependency, position)
        xml.select.bindString(text)
        xml.options.bind(options, GeoAddress()){ (obs) in 
            var xml = ComponentTextXml()
            var view = xml.setup(dependency)
            xml.label.bindString(obs.transformed{ (it) in 
                it.oneLine()
            })
            xml.xmlRoot.onClick{ () in 
                self.position.value = obs.value.coordinate
            }
            return view
        }
        text.addAndRunWeak(self){ (self, value) in 
            delay(1000){ () in 
                if self.text.value == value, value.isNotBlank() {
                    dependency.geocode(value){ (addresses) in 
                        self.options.value = addresses
                    }
                }
            }
        }
        return view
    }
    override public func generate(_ dependency: ViewDependency) -> View {
        return generate(dependency: dependency)
    }
    
    override public init() {
        let text: StandardObservableProperty<String> = StandardObservableProperty("")
        self.text = text
        let position: StandardObservableProperty<GeoCoordinate?> = StandardObservableProperty(nil)
        self.position = position
        let options: StandardObservableProperty<Array<GeoAddress>> = StandardObservableProperty([])
        self.options = options
        super.init()
    }
}
 
